daemon/api_connections.go

```diff
@@ -0,0 +1,229 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
+
+/*
+ * Copyright (C) 2019 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package daemon
+
+import (
+	"net/http"
+	"sort"
+
+	"github.com/snapcore/snapd/interfaces"
+	"github.com/snapcore/snapd/overlord/auth"
+	"github.com/snapcore/snapd/overlord/ifacestate"
+)
+
+var connectionsCmd = &Command{
+	Path:   "/v2/connections",
+	UserOK: true,
+	GET:    getConnections,
+}
+
+type collectFilter struct {
+	snapName  string
+	ifaceName string
+	connected bool
+}
+
+func (c *collectFilter) plugMatches(plug *interfaces.PlugRef, connectedSlots []interfaces.SlotRef) bool {
+	for _, slot := range connectedSlots {
+		if c.slotMatches(&slot, nil) {
+			return true
+		}
+	}
+	if c.snapName != "" && plug.Snap != c.snapName {
+		return false
+	}
+	return true
+}
+
+func (c *collectFilter) slotMatches(slot *interfaces.SlotRef, connectedPlugs []interfaces.PlugRef) bool {
+	for _, plug := range connectedPlugs {
+		if c.plugMatches(&plug, nil) {
+			return true
+		}
+	}
+	if c.snapName != "" && slot.Snap != c.snapName {
+		return false
+	}
+	return true
+}
+
+func (c *collectFilter) ifaceMatches(ifaceName string) bool {
+	if c.ifaceName != "" && c.ifaceName != ifaceName {
+		return false
+	}
+	return true
+}
+
+type bySlotRef []interfaces.SlotRef
+
+func (b bySlotRef) Len() int      { return len(b) }
+func (b bySlotRef) Swap(i, j int) { b[i], b[j] = b[j], b[i] }
```

Comment:
the slot sorting code doesn't seem coverered

Reply:
Added relevant tests now.